File: tree/path_and_sum/path_sum.go
package pathandsum

import "trouble_solver/tree"

// 根节点到叶子节点的路径上所有节点和是否等于targetSum
func hasPathSum(root *tree.TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}
	var curSum int

	var process func(root *tree.TreeNode, targetSum int, curSum int) bool
	process = func(root *tree.TreeNode, targetSum, curSum int) bool {
		if root == nil {
			return false
		}
		curSum += root.Val
		if root.Left == nil && root.Right == nil && curSum == targetSum {
			return true
		}

		return process(root.Left, targetSum, curSum) || process(root.Right, targetSum, curSum)
	}

	return process(root, targetSum, curSum)
}

// 但是实际上，一般有target的题目，我们可以不用定义新变量，通过target==0来判断
func hasPathSumBetter(root *tree.TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}

	if root.Left == nil && root.Right == nil && targetSum-root.Val == 0 {
		return true
	}

	return hasPathSumBetter(root.Left, targetSum-root.Val) || hasPathSumBetter(root.Right, targetSum-root.Val)
}

// 可以看出，这实际上是前序遍历，因此也可以使用栈来非递归写出来

File: tree/path_and_sum/PathSum.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-11-01 19:55:33
 * @LastEditTime: 2023-01-21 14:44:25
 */
package tree.path_and_sum;

import tree.TreeNode;

/**
 * leetcode 112 simple 路径总和
 * 给你二叉树的根节点 root 和一个表示目标和的整数 targetSum 。
 * 判断该树中是否存在 根节点到叶子节点 的路径，这条路径上所有节点值相加等于目标和 targetSum 。
 * 如果存在，返回 true ；否则，返回 false 。
 * 叶子节点 是指没有子节点的节点。
 */
public class PathSum {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }

        if (root.left == null && root.right == null && root.val - targetSum == 0) {
            return true;
        }

        return hasPathSum(root.left, targetSum-root.val) || hasPathSum(root.right, targetSum-root.val);
    }
}
